code/datums/lazy_template.dm

```diff
@@ -46,25 +44,55 @@
 	if(!load_path || !fexists(load_path))
 		CRASH("lazy template [type] has an invalid load_path: '[load_path]', check directory and map name!")
 
-	var/datum/map_template/loading = new(path = load_path, cache = TRUE)
-	if(!loading.cached_map)
+	var/datum/parsed_map/parsed_template = load_map(
+		file(load_path),
+		measure_only = TRUE,
+	)
+	if(isnull(parsed_template.parsed_bounds))
 		CRASH("Failed to cache lazy template for loading: '[key]'")
 
-	var/datum/turf_reservation/reservation = SSmapping.RequestBlockReservation(loading.width, loading.height)
+	var/width = parsed_template.parsed_bounds[MAP_MAXX] - parsed_template.parsed_bounds[MAP_MINX] + 1
+	var/height = parsed_template.parsed_bounds[MAP_MAXY] - parsed_template.parsed_bounds[MAP_MINY] + 1
+	var/datum/turf_reservation/reservation = SSmapping.request_turf_block_reservation(
+		width,
+		height,
+		parsed_template.parsed_bounds[MAP_MAXZ],
+	)
 	if(!reservation)
 		CRASH("Failed to reserve a block for lazy template: '[key]'")
 
-	if(!loading.load(coords2turf(reservation.bottom_left_coords)))
-		CRASH("Failed to load lazy template: '[key]'")
-	reservations += reservation
+	var/list/my_loaded_atoms = list()
+	for(var/z_idx in parsed_template.parsed_bounds[MAP_MAXZ] to 1 step -1)
+		var/turf/bottom_left = reservation.bottom_left_turfs[z_idx]
+		var/turf/top_right = reservation.top_right_turfs[z_idx]
+		load_map(
+			file(load_path),
+			bottom_left.x,
+			bottom_left.y,
+			bottom_left.z,
+			z_upper = z_idx,
+			z_lower = z_idx,
+		)
+		var/list/to_init = list()
+		for(var/turf/turf as anything in block(bottom_left, top_right))
+			to_init |= get_area(turf)
+			to_init |= turf
+			for(var/thing in turf.get_all_contents())
+				to_init |= thing
```

Comment:
at the very least turfs will never duplicate, why use |= for them.
frankly this would be faster if you had a list for areas, a list for turfs and a list for objects, and then +='d them together when you finished.